django/db/models/query.py

```diff
@@ -963,7 +963,27 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+
+            update_fields = set(defaults)
+            concrete_field_names = self.model._meta._non_pk_concrete_field_names
+
+            # `update_fields` does not support non-concrete fields (#31382)
+            if concrete_field_names.issuperset(update_fields):
+                # Add fields which are set on pre_save, e.g. `auto_now` fields.
+                # This is to maintain backward compatibility as these fields are not
+                # updated if not specified explicitly in the `update_fields` list.
+                for field in self.model._meta.local_concrete_fields:
+                    if (
+                        not field.primary_key
+                        and field.__class__.pre_save is not Field.pre_save
+                    ):
```

Comment:
Maybe :thinking: 
```suggestion
                    if not (
                        field.primary_key
                        or field.__class__.pre_save is Field.pre_save
                    ):
```
according to de Morgan's laws.

Reply:
Updated thank you :+1: also rebased so it is just the 1 commit